File: Windows/update-network.js
'use strict';

// Official release-asset API only: no public relay, embedded token or TLS bypass.
function githubReleaseAsset(value) {
  let url;
  try { url = new URL(value); } catch { return null; }
  if (url.protocol !== 'https:' || url.hostname !== 'github.com' || url.port || url.username || url.password || url.search || url.hash) return null;
  const parts = url.pathname.split('/').slice(1);
  if (parts.length !== 6 || parts[2] !== 'releases') return null;
  const [owner, repository] = parts;
  if (![owner, repository].every(part => /^[\w.-]+$/.test(part) && part !== '.' && part !== '..')) return null;
  const latest = parts[3] === 'latest' && parts[4] === 'download';
  if (!latest && parts[3] !== 'download') return null;
  let tag, name;
  try { tag = decodeURIComponent(parts[4]); name = decodeURIComponent(parts[5]); } catch { return null; }
  if (!name || /[\\/\x00-\x1f]/.test(name)) return null;
  const base = `https://api.github.com/repos/${owner}/${repository}/releases`;
  return { repository: `${owner}/${repository}`, name, latest, tag,
    metadataURL: latest ? `${base}/latest` : `${base}/tags/${encodeURIComponent(tag)}`,
    assetPrefix: `${base}/assets/` };
}

function assetAPIURL(source, release) {
  if (!release || release.draft || (!source.latest && release.tag_name !== source.tag)) throw new Error('官方备用通道返回了不匹配的版本。');
  const asset = release.assets?.find(item => item?.name === source.name && item.state === 'uploaded');
  const id = asset?.id;
  if (!Number.isSafeInteger(id) || id <= 0 || asset.url !== `${source.assetPrefix}${id}`) throw new Error('官方备用通道尚未提供此更新文件。');
  return asset.url;
}

class UpdateNetwork {
  constructor({ fetch, userAgent, timeoutMS = 10000 }) {
    this.fetch = fetch;
    this.userAgent = userAgent;
    this.timeoutMS = timeoutMS;
    this.metadata = new Map();
    this.preferAPI = new Set();
  }
  async json(url, accept = 'application/json') {
    const response = await this.fetch(url, {
      signal: AbortSignal.timeout(this.timeoutMS), cache: 'no-store',
      headers: { 'User-Agent': this.userAgent, Accept: accept }
    });
    if (!response.ok) { await response.body?.cancel(); throw new Error(`更新服务返回 HTTP ${response.status}`); }
    return response.json();
  }
  async apiSource(source) {
    let cached = this.metadata.get(source.metadataURL);
    if (!cached || Date.now() - cached.time > 60000) {
      cached = { time: Date.now(), release: await this.json(source.metadataURL, 'application/vnd.github+json') };
      this.metadata.set(source.metadataURL, cached);
    }
    return { url: assetAPIURL(source, cached.release), headers: { Accept: 'application/octet-stream' }, officialAPI: true };
  }
  async withFallback(url, operation) {
    const github = githubReleaseAsset(url);
    const order = github && this.preferAPI.has(github.repository) ? ['api', 'primary'] : ['primary', 'api'];
    let lastError;
    for (const route of order) {
      if (route === 'api' && !github) continue;
      try {
        const source = route === 'api' ? await this.apiSource(github) : { url, headers: {}, officialAPI: false };
        const result = await operation(source);
        if (github) {
          if (route === 'api') this.preferAPI.add(github.repository);
          else this.preferAPI.delete(github.repository);
        }
        return result;
      } catch (error) { lastError = error; }
    }
    throw lastError || new Error('暂时无法连接更新服务器。');
  }
  manifest(url, validate) {
    return this.withFallback(url, async source => validate(await this.json(source.url, source.headers.Accept)));
  }
}
module.exports = { UpdateNetwork, githubReleaseAsset, assetAPIURL };

File: Sources/BiliFetch/UpdateNetwork.swift
import Foundation

struct GitHubReleaseAsset {
    let repository: String
    let name: String
    let tag: String?
    let metadataURL: URL
    let assetPrefix: String

    init?(_ url: URL) {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.scheme == "https", components.host == "github.com",
              components.port == nil, components.user == nil, components.password == nil,
              components.query == nil, components.fragment == nil else { return nil }
        let parts = components.percentEncodedPath.split(separator: "/", omittingEmptySubsequences: false).dropFirst().map(String.init)
        guard parts.count == 6, parts[2] == "releases",
              parts.prefix(2).allSatisfy({ $0.range(of: #"^[A-Za-z0-9_.-]+$"#, options: .regularExpression) != nil && $0 != "." && $0 != ".." }),
              let name = parts[5].removingPercentEncoding, !name.isEmpty,
              name.rangeOfCharacter(from: .controlCharacters.union(CharacterSet(charactersIn: "/\\"))) == nil else { return nil }
        let latest = parts[3] == "latest" && parts[4] == "download"
        guard latest || parts[3] == "download", let decodedTag = parts[4].removingPercentEncoding else { return nil }
        repository = "\(parts[0])/\(parts[1])"
        self.name = name
        tag = latest ? nil : decodedTag
        let base = "https://api.github.com/repos/\(repository)/releases"
        guard let metadataURL = URL(string: latest ? "\(base)/latest" : "\(base)/tags/\(parts[4])") else { return nil }
        self.metadataURL = metadataURL
        assetPrefix = "\(base)/assets/"
    }

    func apiURL(in data: Data) throws -> URL {
        struct Release: Decodable {
            struct Asset: Decodable { let id: Int64; let name: String; let state: String; let url: String }
            let tag_name: String
            let draft: Bool
            let assets: [Asset]
        }
        let release = try JSONDecoder().decode(Release.self, from: data)
        guard !release.draft, tag == nil || tag == release.tag_name,
              let asset = release.assets.first(where: { $0.name == name && $0.state == "uploaded" }),
              asset.id > 0, asset.url == "\(assetPrefix)\(asset.id)", let url = URL(string: asset.url) else {
            throw AppUpdateError.downloadFailed("官方备用通道尚未提供此更新文件。")
        }
        return url
    }
}

@MainActor
final class UpdateNetwork {
    struct Source {
        let url: URL
        let officialAPI: Bool
        func request(userAgent: String, timeout: TimeInterval = 10) -> URLRequest {
            var request = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: timeout)
            request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
            request.setValue(officialAPI ? "application/octet-stream" : "application/json", forHTTPHeaderField: "Accept")
            return request
        }
    }
    let userAgent: String
    private let session: URLSession
    private var metadata: [URL: (date: Date, data: Data)] = [:]
    private var preferredAPI = Set<String>()
    init(userAgent: String, session: URLSession = .shared) { self.userAgent = userAgent; self.session = session }

    private func data(for request: URLRequest) async throws -> Data {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw AppUpdateError.downloadFailed("更新服务器没有返回有效数据。")
        }
        return data
    }
    private func apiSource(_ github: GitHubReleaseAsset) async throws -> Source {
        let cached = metadata[github.metadataURL]
        let releaseData: Data
        if let cached, Date().timeIntervalSince(cached.date) < 60 {
            releaseData = cached.data
        } else {
            var request = Source(url: github.metadataURL, officialAPI: false).request(userAgent: userAgent)
            request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
            releaseData = try await data(for: request)
            metadata[github.metadataURL] = (Date(), releaseData)
        }
        return Source(url: try github.apiURL(in: releaseData), officialAPI: true)
    }
    func withFallback<T>(_ url: URL, operation: (Source) async throws -> T) async throws -> T {
        let github = GitHubReleaseAsset(url)
        let order = github.map { preferredAPI.contains($0.repository) } == true ? [true, false] : [false, true]
        var lastError: Error = AppUpdateError.downloadFailed("暂时无法连接更新服务器。")
        for useAPI in order {
            if useAPI && github == nil { continue }
            do {
                let source: Source
                if useAPI, let github { source = try await apiSource(github) }
                else { source = Source(url: url, officialAPI: false) }
                let result = try await operation(source)
                if let github {
                    if useAPI { preferredAPI.insert(github.repository) }
                    else { preferredAPI.remove(github.repository) }
                }
                return result
            } catch { lastError = error }
        }
        throw lastError
    }
    func manifest(_ url: URL, currentVersion: String) async throws -> AppUpdateRelease {
        try await withFallback(url) { source in
            let content = try await self.data(for: source.request(userAgent: self.userAgent))
            return try JSONDecoder().decode(AppUpdateManifest.self, from: content).macOSRelease(currentVersion: currentVersion)
        }
    }
}
